nmigen_boards/ecp5_5g_evn.py

```diff
@@ -15,20 +15,117 @@ class ECP55GEVNPlatform(LatticeECP5Platform):
+        assert VCCIO2 in ["LVCMOS33", "LVCMOS25"]
+        assert VCCIO6 in ["LVCMOS33", "LVCMOS25"]
+        assert VCCIO7 in ["LVCMOS33", "LVCMOS25", "LVCMOS15", "LVCMOS12"]
+        self._VCCIO1 = VCCIO1
+        self._VCCIO2 = VCCIO2
+        self._VCCIO6 = VCCIO6
+        self._VCCIO7 = VCCIO7
+    
+    def VCCIO1(self):
+        return self._VCCIO1
+    
+    def VCCIO2(self):
+        return self._VCCIO2
+    
+    def VCCIO6(self):
+        return self._VCCIO6
+    
+    def VCCIO7(self):
+        return self._VCCIO7
+    
     resources   = [
         Resource("rst", 0, PinsN("G2", dir="i"), Attrs(IO_TYPE="LVCMOS33")),
         Resource("clk12", 0, Pins("A10", dir="i"),
                  Clock(12e6), Attrs(IO_TYPE="LVCMOS33")),
-
-        *LEDResources(pins="A13 A12 B19 A18 B18 C17 A17 B17",
-                      attrs=Attrs(IO_TYPE="LVCMOS33")),
-        *SwitchResources(pins="P4", invert=True,
-                         attrs=Attrs(IO_TYPE="LVCMOS33")),
-
+        
+        # The switches and LEDs are connected over resistors to VCCIO6 for SW4 and SW5 switch 1-3 and VCCIO1 for the LEDs and SW5 switch 4-8.
+        # VCCIO1 is connected by default to 2.5 V over R100 (can be set to 3.3 V by disconnecting R100 and connecting R105)
+        # VCCIO6 is connected to 3.3 V by default over R99 (can be switched to 2.5 V with R104, see page 51 in the ECP5 evaluation board datasheet)
+        *LEDResources(pins="A13 A12 B19 A18 B18 C17 A17 B17", invert=True,
+                      attrs=Attrs(IO_TYPE=VCCIO1)),
+        
+        *ButtonResources(pins="P4", invert=True,
+                         attrs=Attrs(IO_TYPE=VCCIO6)),
+        
+        *SwitchResources(pins={1: "J1", 2: "H1", 3: "K1"}, invert=True,
+                         attrs=Attrs(IO_TYPE=VCCIO6)),
+        
+        *SwitchResources(pins={4: "E15", 5: "D16", 6: "B16", 7: "C16", 8: "A16"}, invert=True,
+                         attrs=Attrs(IO_TYPE=VCCIO1)),
+        
+        # SERDES
+        Resource("serdes", 0,
+            Subsignal("tx", DiffPairs("W4", "W5", dir="o")),
+            Subsignal("rx", DiffPairs("Y5", "Y6", dir="i")),
+        ),
+        Resource("serdes", 1,
+            Subsignal("tx", DiffPairs("W8", "W9", dir="o")),
+            Subsignal("rx", DiffPairs("Y7", "Y8", dir="i")),
+        ),
+        Resource("serdes", 2,
+            Subsignal("tx", DiffPairs("W13", "W14", dir="o")),
+            Subsignal("rx", DiffPairs("Y14", "Y15", dir="i")),
+        ),
+        Resource("serdes", 3,
+            Subsignal("tx", DiffPairs("W17", "W18", dir="o")),
+            Subsignal("rx", DiffPairs("Y16", "Y17", dir="i")),
+        ),
+        
+        Resource("serdesclk", 0, DiffPairs("Y11", "Y12", dir="o")),
+        Resource("serdesclk", 1, DiffPairs("Y19", "W20", dir="o")), # 200 MHz
+        
+        # Diff pairs, they have selectable termination resistors
+        # Resource("diff", "0", DiffPairs(A4, A5), Attrs(IO_TYPE=VCCIO7)),
```

Comment:
I apologize for making you do extra work because I misremembered how this board works. While the SERDES pins should indeed be described as resources (since every application will have to add the same resources anyway), the diffpairs should not be (since every application will use an application-specific resource incorporating these).